### PR TITLE
SNOW-878366 Don't require user or password for particular authenticators

### PR DESCRIPTION
**Pull request by sfc-gh-pfus**
Provided that only when authenticator really requires user/password this properties are required.

### CODE REVIEW

**Review by sfc-gh-pbulawa — APPROVED** [review]

**Review by sfc-gh-dprzybysz — APPROVED** [review]

File: Snowflake.Data/Core/Session/SFSessionProperty.cs [hunk 1]
```diff
@@ -292,23 +292,28 @@ private static bool IsRequired(SFSessionProperty sessionProperty, SFSessionPrope
                 var authenticatorDefined =
                     properties.TryGetValue(SFSessionProperty.AUTHENTICATOR, out var authenticator);
 
+                List<string> authenticatorsWithoutPassword = new List<string>()
```
Comment (sfc-gh-knozderko) [inline Snowflake.Data/Core/Session/SFSessionProperty.cs:295]: you could use var here

File: Snowflake.Data/Core/Session/SFSessionProperty.cs [hunk 2]
```diff
@@ -292,23 +292,28 @@ private static bool IsRequired(SFSessionProperty sessionProperty, SFSessionPrope
...
             }
             else if (sessionProperty.Equals(SFSessionProperty.USER))
             {
                 var authenticatorDefined =
                    properties.TryGetValue(SFSessionProperty.AUTHENTICATOR, out var authenticator);
 
-                // Oauth don't require a username for authenticating
-                return !(authenticatorDefined && (
-                    authenticator.Equals(OAuthAuthenticator.AUTH_NAME, StringComparison.OrdinalIgnoreCase)));
+                List<string> authenticatorsWithoutUsername = new List<string>()
```
Comment (sfc-gh-knozderko) [inline Snowflake.Data/Core/Session/SFSessionProperty.cs:310]: use var

File: Snowflake.Data.Tests/SFSessionPropertyTest.cs [hunk 3]
```diff
@@ -151,7 +151,7 @@ public static IEnumerable<TestCase> ConnectionStringTestCases()
                 simpleTestCase,
                 testCaseWithBrowserResponseTimeout,
                 testCaseWithProxySettings,
-                testCaseThatDefaultForUseProxyIsFalse
+                testCaseThatDefaultForUseProxyIsFalse,
```
Comment (sfc-gh-knozderko) [inline Snowflake.Data.Tests/SFSessionPropertyTest.cs:154]: is the comma at the end really needed?

File: README.md [hunk 4]
```diff
@@ -290,7 +289,8 @@ If you are using a different method for authentication, see the examples below:
 
 * **Browser-based SSO**
 
-  In the connection string, set `AUTHENTICATOR=externalbrowser`, and set `USER` to the login name for your IdP.
+  In the connection string, set `AUTHENTICATOR=externalbrowser`.
+  Optionally, `USER` can set set. In that case only if user authenticated via external browser matches the one from configuration, authentication will complete.
```
Comment (sfc-gh-knozderko) [inline README.md:293]: set set?

**Review by sfc-gh-knozderko — CHANGES_REQUESTED** [review]

**Review by sfc-gh-knozderko — APPROVED** [review]

**Pull request merged**